File: src/main/java/pl/miloszgilga/command/vote/VoteSkipQueueToTrackCmd.java
/*
 * Copyright (c) 2023 by MILOSZ GILGA <http://miloszgilga.pl>
 *
 * File name: VoteSkipQueueToTrackCmd.java
 * Last modified: 17/05/2023, 01:30
 * Project name: jwizard-discord-bot
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     <http://www.apache.org/license/LICENSE-2.0>
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the license.
 */

package pl.miloszgilga.command.vote;

import net.dv8tion.jda.api.entities.Guild;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;

import java.util.Map;

import pl.miloszgilga.BotCommand;
import pl.miloszgilga.BotCommandArgument;
import pl.miloszgilga.misc.Utilities;
import pl.miloszgilga.locale.ResLocaleSet;
import pl.miloszgilga.dto.CommandEventWrapper;
import pl.miloszgilga.vote.VoteEmbedResponse;
import pl.miloszgilga.audioplayer.MusicManager;
import pl.miloszgilga.audioplayer.PlayerManager;
import pl.miloszgilga.embed.EmbedMessageBuilder;
import pl.miloszgilga.command.AbstractVoteMusicCommand;
import pl.miloszgilga.cacheable.CacheableCommandStateDao;
import pl.miloszgilga.core.remote.RemotePropertyHandler;
import pl.miloszgilga.core.configuration.BotConfiguration;
import pl.miloszgilga.core.loader.JDAInjectableCommandLazyService;

import static pl.miloszgilga.exception.AudioPlayerException.TrackQueueIsEmptyException;
import static pl.miloszgilga.exception.AudioPlayerException.TrackPositionOutOfBoundsException;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@JDAInjectableCommandLazyService
public class VoteSkipQueueToTrackCmd extends AbstractVoteMusicCommand {

    VoteSkipQueueToTrackCmd(
        BotConfiguration config, PlayerManager playerManager, EmbedMessageBuilder embedBuilder,
        RemotePropertyHandler handler, CacheableCommandStateDao cacheableCommandStateDao
    ) {
        super(BotCommand.VOTE_SKIP_TO_TRACK, config, playerManager, embedBuilder, handler, cacheableCommandStateDao);
        super.onSameChannelWithBot = true;
        super.inPlayingMode = true;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    protected VoteEmbedResponse doExecuteVoteMusicCommand(CommandEventWrapper event) {
        final MusicManager musicManager = playerManager.getMusicManager(event);
        final Integer trackPos = event.getArgumentAndParse(BotCommandArgument.VOTE_SKIP_TRACK_POSITION);

        if (musicManager.getQueue().isEmpty()) {
            throw new TrackQueueIsEmptyException(config, event);
        }
        if (musicManager.getActions().checkInvTrackPosition(trackPos)) {
            throw new TrackPositionOutOfBoundsException(config, event, musicManager.getQueue().size());
        }
        final AudioTrack currentPlaying = musicManager.getAudioPlayer().getPlayingTrack();
        final AudioTrack trackToSkipped = musicManager.getActions().getTrackByPosition(trackPos);

        final Map<String, Object> attributes = Map.of(
            "audioTrack", Utilities.getRichTrackTitle(currentPlaying.getInfo()),
            "nextAudioTrack", Utilities.getRichTrackTitle(trackToSkipped.getInfo()),
            "countOfSkipped", String.valueOf(trackPos - 1)
        );
        final Guild g = event.getGuild();
        return new VoteEmbedResponse(
            VoteSkipQueueToTrackCmd.class,
            embedBuilder.createInitialVoteMessage(event, ResLocaleSet.VOTE_SKIP_TO_TRACK_MESS, attributes),
            ed -> {
                playerManager.skipToTrackPos(event, trackPos);
                return embedBuilder.createSuccessVoteMessage(ResLocaleSet.SUCCESS_VOTE_SKIP_TRACK_MESS, attributes, ed, g);
            },
            ed -> embedBuilder.createFailureVoteMessage(ResLocaleSet.FAILURE_VOTE_SKIP_TO_TRACK_MESS, attributes, ed, g),
            ed -> embedBuilder.createTimeoutVoteMessage(ResLocaleSet.FAILURE_VOTE_SKIP_TO_TRACK_MESS, attributes, ed, g)
        );
    }
}

File: src/main/kotlin/pl/jwizard/core/api/vote/VoteSkipQueueToTrackCmd.kt
/*
 * Copyright (c) 2024 by JWizard
 * Originally developed by Miłosz Gilga <https://miloszgilga.pl>
 */
package pl.jwizard.core.api.vote

import pl.jwizard.core.api.AbstractVoteMusicCmd
import pl.jwizard.core.audio.player.PlayerManagerFacade
import pl.jwizard.core.bot.BotConfiguration
import pl.jwizard.core.command.BotCommand
import pl.jwizard.core.command.CompoundCommandEvent
import pl.jwizard.core.command.arg.CommandArgument
import pl.jwizard.core.command.reflect.CommandListenerBean
import pl.jwizard.core.exception.AudioPlayerException
import pl.jwizard.core.i18n.I18nResLocale
import pl.jwizard.core.util.Formatter
import pl.jwizard.core.vote.VoteResponseData

@CommandListenerBean(id = BotCommand.VSKIPTO)
class VoteSkipQueueToTrackCmd(
	botConfiguration: BotConfiguration,
	playerManagerFacade: PlayerManagerFacade
) : AbstractVoteMusicCmd(
	botConfiguration,
	playerManagerFacade
) {
	init {
		onSameChannelWithBot = true
		inPlayingMode = true
	}

	override fun executeVoteMusicCmd(event: CompoundCommandEvent): VoteResponseData {
		val musicManager = playerManagerFacade.findMusicManager(event)
		if (musicManager.queue.isEmpty()) {
			throw AudioPlayerException.TrackQueueIsEmptyException(event)
		}
		val trackPosition = getArg<Int>(CommandArgument.POS, event)
		if (musicManager.actions.checkInvertedTrackPosition(trackPosition)) {
			throw AudioPlayerException.TrackPositionOutOfBoundsException(event, musicManager.queue.size)
		}
		val currentPlaying = musicManager.audioPlayer.playingTrack
		val trackToSkipped = musicManager.actions.getTrackByPosition(trackPosition)
		val params = mapOf(
			"audioTrack" to Formatter.createRichTrackTitle(currentPlaying),
			"nextAudioTrack" to Formatter.createRichTrackTitle(trackToSkipped),
			"countOfSkipped" to (trackPosition - 1).toString()
		)
		return VoteResponseData(
			initClazz = VoteSkipQueueToTrackCmd::class,
			message = buildInitMessage(I18nResLocale.VOTE_SKIP_TO_TRACK, params, event),
			onSuccess = {
				playerManagerFacade.skipToTrackPos(event, trackPosition)
				buildSuccessMessage(I18nResLocale.SUCCESS_VOTE_SKIP_TO_TRACK, params, it, event)
			},
			onFailure = { buildFailureMessage(I18nResLocale.FAILURE_VOTE_SKIP_TO_TRACK, params, it, event) },
			onTimeout = { buildTimeoutMessage(I18nResLocale.FAILURE_VOTE_SKIP_TO_TRACK, params, it, event) },
		)
	}
}
